mainapp/views/views.py

```diff
@@ -243,18 +247,51 @@ def file_serve_proxy(
 
 
 def file_serve(request, id):
-    logger.warning("Serving media files through django is slow")
-
     """ Ensure that the file is not deleted in the database """
-    get_object_or_404(File, id=id)
+    file = get_object_or_404(File, id=id)
 
-    minio_file = minio_client().get_object(minio_file_bucket, str(id))
-    response = HttpResponse(minio_file.read())
+    name, ext = splitext(file.filename)
+    if name.isnumeric() and file.name:
+        filename = f'{file.name}_{name}{ext}'
+    else:
+        filename = file.filename
 
-    response["Content-Type"] = minio_file.headers["Content-Type"]
+    headers = {
+        "Content-Disposition": f'attachment; filename="{filename}"',
+        "Content-Type": str(file.mime_type),
+    }
 
     if settings.SITE_SEO_NOINDEX:
-        response["X-Robots-Tag"] = "noindex"
+        headers["X-Robots-Tag"] = "noindex"
+
+    if settings.MINIO_REDIRECT:
+        if settings.MINIO_PUBLIC_HOST:
+            endpoint = settings.MINIO_PUBLIC_HOST
+            secure = settings.MINIO_PUBLIC_SECURE
+        else:
+            endpoint = settings.MINIO_HOST
+            secure = settings.MINIO_SECURE
+
+        mc = minio_client()
+        old_base_url = mc._base_url
+        mc._base_url = minio.api.BaseURL(("https://" if secure else "http://") + endpoint, settings.MINIO_REGION)
```

Comment:
please avoid using private variables. in this case you can pass the settings on the `minio_client` function or just construct a new minio client object in the spot

Reply:
Agreed. That was a nasty workaround.

I fixed it by creating two Minio instances of which one uses the public S3 endpoint of configured with MINIO_PUBLIC_HOST. This instance is only used for generating the presigned URLs. All the actual S3 operations are targeting always the default endpoint.